### PR TITLE
fix oai proxy

### PR DESCRIPTION
**Pull request by rhjdvsgsgks**
fix generation not stoped while bot stop talking in chat mode

fix possible `slot_id` not exist

response for cors (and pre flight)

### CODE REVIEW

**Comment by S-Ali-Zaidi** [comment]
> Hi, thank you for implementing this fix! It looks like a lot of what I tired to tinker with myself is what you ended up doing here too.
> 
> One issue that is still on-going: the api_like_oai.py proxy still doesn’t seem compatible with all OpenAI served LLM chat UI clients. For instance, I’ve included a screenshot here that shows the following:
> 
> 1. Top-middle: Me running the api_like_oai.py at port 8081.
> 2. Bottom-middle: Me running successful curl requests to the proxy server at port 8081
> 3. Top-right and bottom-right: Using the MacGPT client, redirected to port 8081, to successfully use the llama.cpp endpoint instead of OpenAI endpoints.
> 4. Left: using the Chatbox UI, also directed to port 8081, to also send requests to the llama.cpp server -- but constantly getting back empty / non-generative streams of text.. despite the fact that I see the POST requests going through successful on the terminal screen for api_like_OAI AND seeing the text being received, generated, and sent back on the terminal screen for ./server (not pictured here).
> 
> Screenshot 2023-11-07 at 2 52 16 PM
> I should note that I’ve tested Chatbox with the actual OpenAI servers **as well as other llama.cpp based servers, such as Studio LM*** and Chatbox works perfectly fine with those endpoints and streams back text without issue.
> 
> I’ve been trying for a few days to figure out how to fix this and I’m totally stuck. Do you have any insights on how to fix this with the api_like_oai.py script?
> 
> To recap:
> 
> * api_like_oai.py does not work with all UI clients designed to interface with OpenAI endpoints, such as Chatbox. The issue is receiving back streamed text that is understood and rendered by the Chatbox client.
> * api_like_oai.py is working with curl and with some other chat clients like MacGPT.
> * Chatbox has been tested and confirmed to be working with the openAI servers, and to be working when sending requests to port 8081 if the llama.cpp framework is being served by other programs, such as LM Studio.


Oh geez, never mind, I answered my own question, finally, after days of trying to figure this out... 

There MUST be TWO newline breaks in the JSON returns between each data chunk in order to ensure that all UI clients can understand and parse the chunks correctly. 

So it needs to read `\n\n` instead of `\n` in the lines `yield 'data: {}\n\n'.format(json.dumps(resData))` as follows in the `def generate():` function:

```
def generate():
            data = requests.request("POST", urllib.parse.urljoin(args.llama_api, "/completion"), data=json.dumps(postData), stream=True)
            time_now = int(time.time())
            resData = make_resData_stream({}, chat=True, time_now=time_now, start=True)
            yield 'data: {}\n\n'.format(json.dumps(resData))
            for line in data.iter_lines():
                if line:
                    decoded_line = line.decode('utf-8')
                    resData = make_resData_stream(json.loads(decoded_line[6:]), chat=True, time_now=time_now)
                    yield 'data: {}\n\n'.format(json.dumps(resData))
```

This is a super critical thing to add, otherwise many clients will not render the text stream! 

I’m very new to all of this.. so unsure how to make the modification myself to submit as a potential commit. Hoping someone else can?

File: examples/server/api_like_OAI.py [hunk 1]
```diff
@@ -217,7 +221,7 @@ def generate():
                     decoded_line = line.decode('utf-8')
                     resData = make_resData_stream(json.loads(decoded_line[6:]), chat=False, time_now=time_now)
                     yield 'data: {}\n'.format(json.dumps(resData))
```
Comment (S-Ali-Zaidi) [inline examples/server/api_like_OAI.py:223]: There need to be TWO `\n` to ensure that all chat clients can understand the streamed data chunks. 

Should read `yield 'data: {}\n\n'.format(json.dumps(resData))`

File: examples/server/api_like_OAI.py [hunk 2]
```diff
@@ -183,14 +185,16 @@ def generate():
                     decoded_line = line.decode('utf-8')
                     resData = make_resData_stream(json.loads(decoded_line[6:]), chat=True, time_now=time_now)
                     yield 'data: {}\n'.format(json.dumps(resData))
```
Comment (S-Ali-Zaidi) [inline examples/server/api_like_OAI.py:187]: There need to be TWO `\n` to ensure that all chat clients can understand the streamed data chunks. 

Should read `yield 'data: {}\n\n'.format(json.dumps(resData))`

**Review by zaidifm — CHANGES_REQUESTED** [review]
There need to be TWO `\n` to ensure that all chat clients can understand the streamed data chunks. 

relevant lines the generate function read `yield 'data: {}\n\n'.format(json.dumps(resData))`

**Comment by Azeirah** [comment]
The fix with the two required `\n`'s has been proposed before here https://github.com/ggerganov/llama.cpp/pull/2383

It might be worth checking that PR as well.

**Review by zaidifm — APPROVED** [review]
Works great!

**Pull request merged**